### ISSUE TITLE
[clang][enum_overflow]Clang incorrectly thinks that enumerators of enum does not fit in the largest integer type

### ISSUE THREAD

**Issue by zhaojiangkun-1** [llvm/llvm-project#120350]
Test case:
```c
enum E {
  e1 = 0xffffffffffffffff,
  e2, e3
} e = e3;

#define SA(I,X) int a##I[(X)? 1 : -1]

SA(1, sizeof(E) == sizeof(__int128));

```
Backtrace:
```
warning: incremented enumerator value 18446744073709551616 is not representable in the largest integer type [-Wenum-too-large]
  e2, e3
  ^
./ext1/int128-3.C:11:1: error: 'a1' declared as an array with a negative size
SA(1, sizeof(E) == sizeof(__int128));
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./ext1/int128-3.C:9:26: note: expanded from macro 'SA'
#define SA(I,X) int a##I[(X)? 1 : -1]
                         ^~~~~~~~~~~
1 warning and 1 error generated.

```
It can be compiled in gcc, but an error is reported in clang.

https://godbolt.org/z/4Pdxs5sEs

This problem has been fixed before, but is it not compatible with this yet?
https://github.com/llvm/llvm-project/issues/24667

**Comment by hstk30-hw** [comment]
C99 6.7.2.2p2

> The expression that defines the value of an enumeration constant shall be an integer constant expression that has a value representable as an int

C99 6.7.2.2p4 

> Each enumerated type shall be compatible with char, a signed integer type, or an
unsigned integer type. The choice of type is implementation-defined,108) but shall be
capable of representing the values of all the members of the enumeration.


https://github.com/llvm/llvm-project/blob/060d62b48aeb5080ffcae1dc56e41a06c6f56701/clang/lib/Sema/SemaDecl.cpp#L19700-L19711
